### PR TITLE
Add light version of prepare_dump API

### PR DESCRIPTION
**Pull request by tovrstra**
This fixes one task in #191. It introduces a partial version of the `prepare_dump` API proposed in the issue.

This light version of the API only checks the validity of the `IOData` instance, but does not support any conversion yet. (This will be added in future PRs.)

Other changes:

- One unrelated fix (correction of a type hint in `iodata/basis.py`)
- Cleaned up auxiliary function in `test_wfn.py` and `test_wfx.py`, needed for new tests

I will YOLO-merge this on Wednesday June 26, unless reviewed earlier.

## Summary by Sourcery

This pull request introduces a light version of the `prepare_dump` API to validate `IOData` objects' compatibility with different file formats. It also enhances error handling by introducing a specific `PrepareDumpError` and updates the `dump_one` and `dump_many` functions to use this new API. Additionally, it includes tests to ensure the new functionality works as expected.

- **New Features**:
    - Introduced a light version of the `prepare_dump` API to check the compatibility of `IOData` objects with various file formats without supporting any conversion.
- **Bug Fixes**:
    - Corrected a type hint in `iodata/basis.py`.
- **Enhancements**:
    - Replaced `ValueError` with `PrepareDumpError` for more specific error handling in file format compatibility checks.
    - Added `prepare_dump` checks in `dump_one` and `dump_many` functions to ensure `IOData` objects are compatible with the target file formats before dumping.
- **Tests**:
    - Added tests to verify that `prepare_dump` raises `PrepareDumpError` for unsupported generalized molecular orbitals in various formats.

### CODE REVIEW

**Comment by tovrstra** [comment]
The deepsource issues are related to the complexity of some `dump_one` functions. This problem is not due to the current PR, but it made small modifications in such functions, which triggered deepsource.

File: iodata/test/test_fchk.py [hunk 1]
```diff
@@ -690,3 +696,10 @@ def test_dump_fchk_from_wfx_lih_cisd_cation(tmpdir):
 )
 def test_dump_fchk_from_molekel(tmpdir, path, match):
     check_load_dump_consistency(tmpdir, path, match)
+
+
+def test_generalized():
+    # The FCHK format does not support generalized MOs
+    data = create_generalized()
+    with pytest.raises(PrepareDumpError):
+        dump_one(data, "generlized.fchk")
```
Comment (PaulWAyers) [inline iodata/test/test_fchk.py:705]: spelling error; `generalized.fchk`

File: iodata/test/test_molekel.py [hunk 2]
```diff
@@ -157,3 +163,10 @@ def test_load_mkl_h2_huge_threshold():
         warnings.simplefilter("error")
         # The threshold is set very high, which skip a correction for ORCA.
         load_one(str(fn_molekel), norm_threshold=1e4)
+
+
+def test_generalized():
+    # The Molden format does not support generalized MOs
+    data = create_generalized()
+    with pytest.raises(PrepareDumpError):
+        dump_one(data, "generlized.mkl")
```
Comment (PaulWAyers) [inline iodata/test/test_molekel.py:172]: `generalized.mkl` unless there was a reason to keep it to 10 characters or less.

File: iodata/test/test_wfn.py [hunk 3]
```diff
@@ -359,8 +357,20 @@ def test_load_dump_consistency(path, tmpdir):
...
+def test_dump_one_pure_functions(tmpdir):
+    with pytest.raises(PrepareDumpError):
+        check_load_dump_consistency("water_ccpvdz_pure_hf_g03.fchk", tmpdir)
+
+
+def test_generalized():
+    # The WFN format does not support generalized MOs
+    data = create_generalized()
+    with pytest.raises(PrepareDumpError):
+        dump_one(data, "generlized.wfn")
```
Comment (PaulWAyers) [inline iodata/test/test_wfn.py:376]: `generalized.wfn`

File: iodata/test/test_wfx.py [hunk 4]
```diff
@@ -812,3 +814,10 @@ def test_load_one_lih_cation_rohf():
     olp = compute_overlap(mol.obasis, mol.atcoords)
     check_orthonormal(mol.mo.coeffsa, olp, 1e-5)
     check_orthonormal(mol.mo.coeffsb, olp, 1e-5)
+
+
+def test_generalized():
+    # The Molden format does not support generalized MOs
+    data = create_generalized()
+    with pytest.raises(PrepareDumpError):
+        dump_one(data, "generlized.wfx")
```
Comment (PaulWAyers) [inline iodata/test/test_wfx.py:823]: `generalized.wfx`

File: iodata/formats/wfx.py [hunk 5]
```diff
@@ -327,17 +327,34 @@ def load_one(lit: LineIterator) -> dict:
...
+        raise PrepareDumpError("The WFX format requires molecular orbitals.")
+    if data.obasis is None:
+        raise PrepareDumpError("The WFX format requires an orbital basis set.")
+    if data.mo.kind == "generalized":
+        raise PrepareDumpError("Cannot write WFX file with generalized orbitals.")
+    if data.mo.occs_aminusb is not None:
+        raise PrepareDumpError("Cannot write WFX file when mo.occs_aminusb is set.")
+    for shell in data.obasis.shells:
+        if any(kind != "c" for kind in shell.kinds):
+            raise PrepareDumpError("The WFX only supports Cartesian MolecularBasis.")
```
Comment (PaulWAyers) [inline iodata/formats/wfx.py:348]: When we try to dump a `wfx` with spherical, should we convert spherical -> Cartesian and then dump. That's what I think most codes that dump `wfx` do internally.

File: iodata/formats/wfn.py [hunk 6]
```diff
@@ -494,23 +494,39 @@ def _dump_helper_section(f: TextIO, data: NDArray, fmt: str, skip: int, step: in
...
+        raise PrepareDumpError("The WFN format requires molecular orbitals")
+    if data.obasis is None:
+        raise PrepareDumpError("The WFN format requires an orbital basis set")
+    if data.mo.kind == "generalized":
+        raise PrepareDumpError("Cannot write WFN file with generalized orbitals.")
+    if data.mo.occs_aminusb is not None:
+        raise PrepareDumpError("Cannot write WFN file when mo.occs_aminusb is set.")
+    for shell in data.obasis.shells:
+        if any(kind != "c" for kind in shell.kinds):
+            raise PrepareDumpError("The WFN only supports Cartesian MolecularBasis.")
```
Comment (PaulWAyers) [inline iodata/formats/wfn.py:515]: When we try to dump a `wfn` with spherical, should we convert spherical -> Cartesian and then dump. That's what I think most codes that dump `wfn` do internally.

**Review by PaulWAyers — APPROVED** [review]
Minor comments, nothing serious or essential.

File: iodata/test/test_fchk.py (hunk 1 above)
Comment (tovrstra) [reply iodata/test/test_fchk.py]: fixed

File: iodata/test/test_wfn.py (hunk 3 above)
Comment (tovrstra) [reply iodata/test/test_wfn.py]: fixed

File: iodata/test/test_wfx.py (hunk 4 above)
Comment (tovrstra) [reply iodata/test/test_wfx.py]: fixed

File: iodata/test/test_molekel.py (hunk 2 above)
Comment (tovrstra) [reply iodata/test/test_molekel.py]: fixed

File: iodata/formats/wfx.py (hunk 5 above)
Comment (tovrstra) [reply iodata/formats/wfx.py]: I agree. That's part of the plan in #191. I'm aiming to fix this in one of the following PRs. This exception is a not meant to stay like this.

File: iodata/formats/wfn.py (hunk 6 above)
Comment (tovrstra) [reply iodata/formats/wfn.py]: Agreed. Idem. Will be fixed in a not-so distant future PR.

**Comment by tovrstra** [comment]
Thanks for reviewing, Paul!

**Pull request merged**